synapse/storage/databases/main/stream.py

```diff
@@ -357,6 +357,24 @@ def filter_to_clause(event_filter: Optional[Filter]) -> Tuple[str, List[str]]:
         )
         args.extend(event_filter.related_by_rel_types)
 
+    if event_filter.rel_types:
+        clauses.append(
+            "(%s)"
+            % " OR ".join(
+                "event_relation.relation_type = ?" for _ in event_filter.rel_types
+            )
+        )
+        args.extend(event_filter.rel_types)
+
+    if event_filter.not_rel_types:
+        clauses.append(
+            "((%s) OR event_relation.relation_type IS NULL)"
+            % " AND ".join(
+                "event_relation.relation_type != ?" for _ in event_filter.not_rel_types
+            )
+        )
+        args.extend(event_filter.not_rel_types)
```

Comment:
I think this is fine for now, but I am concerned about the query performance of these fields. In particular, I think for uncommon relation types this could cause us to scan the full events table looking for matches?

Reply:
Are you concerned specifically for the `not_rel_types` case or for both cases?

Reply:
I ran a couple of queries against Matrix HQ, one for not `m.thread` and one for not `invalid-rel-type`. Both have similar (identical?) query plans:

<details>
<summary>`m.thread` case</summary>

```sql
=>EXPLAIN ANALYZE SELECT
->     event.event_id, event.instance_name,
->     event.topological_ordering, event.stream_ordering
-> FROM events AS event
-> LEFT JOIN event_relations AS event_relation USING (event_id)
-> WHERE
->     event.outlier = false
->     AND event.room_id = '!OGEhHVWSdvArJzumhm:matrix.org'
->     AND ((505473,3359746309) < (event.topological_ordering,event.stream_ordering))
->     AND ((event_relation.relation_type != 'm.thread') OR event_relation.relation_type IS NULL)
-> ORDER BY event.topological_ordering DESC,
-> event.stream_ordering DESC LIMIT 20;
                                                                             QUERY PLAN                                                                              
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1.27..134.20 rows=20 width=73) (actual time=0.775..6.941 rows=20 loops=1)
   ->  Nested Loop Left Join  (cost=1.27..154360.09 rows=23224 width=73) (actual time=0.774..6.937 rows=20 loops=1)
         Filter: ((event_relation.relation_type <> 'm.thread'::text) OR (event_relation.relation_type IS NULL))
         ->  Index Scan Backward using events_order_room on events event  (cost=0.70..48115.51 rows=23288 width=73) (actual time=0.304..0.500 rows=20 loops=1)
               Index Cond: ((room_id = '!OGEhHVWSdvArJzumhm:matrix.org'::text) AND (ROW(505473, '3359746309'::bigint) < ROW(topological_ordering, stream_ordering)))
               Filter: (NOT outlier)
         ->  Index Scan using event_relations_id on event_relations event_relation  (cost=0.57..4.55 rows=1 width=55) (actual time=0.320..0.320 rows=0 loops=20)
               Index Cond: (event.event_id = event_id)
 Planning time: 0.281 ms
 Execution time: 6.974 ms
(10 rows)
```

</details>

<details>
<summary>`invalid-rel-case` case</summary>

```sql
=> EXPLAIN ANALYZE SELECT
    event.event_id, event.instance_name,
    event.topological_ordering, event.stream_ordering
FROM events AS event
LEFT JOIN event_relations AS event_relation USING (event_id)
WHERE
    event.outlier = false
    AND event.room_id = '!OGEhHVWSdvArJzumhm:matrix.org'
    AND ((505473,3359746309) < (event.topological_ordering,event.stream_ordering))
    AND ((event_relation.relation_type != 'invalid-rel-type') OR event_relation.relation_type IS NULL)
ORDER BY event.topological_ordering DESC,
event.stream_ordering DESC LIMIT 20;
                                                                             QUERY PLAN                                                                              
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1.27..134.20 rows=20 width=73) (actual time=0.069..0.362 rows=20 loops=1)
   ->  Nested Loop Left Join  (cost=1.27..154360.09 rows=23224 width=73) (actual time=0.068..0.356 rows=20 loops=1)
         Filter: ((event_relation.relation_type <> 'invalid-rel-type'::text) OR (event_relation.relation_type IS NULL))
         ->  Index Scan Backward using events_order_room on events event  (cost=0.70..48115.51 rows=23288 width=73) (actual time=0.050..0.105 rows=20 loops=1)
               Index Cond: ((room_id = '!OGEhHVWSdvArJzumhm:matrix.org'::text) AND (ROW(505473, '3359746309'::bigint) < ROW(topological_ordering, stream_ordering)))
               Filter: (NOT outlier)
         ->  Index Scan using event_relations_id on event_relations event_relation  (cost=0.57..4.55 rows=1 width=55) (actual time=0.011..0.011 rows=0 loops=20)
               Index Cond: (event.event_id = event_id)
 Planning time: 0.652 ms
 Execution time: 0.425 ms
(10 rows)
```

</details>